haystack/evaluation/eval.py

```diff
@@ -0,0 +1,123 @@
+from typing import Any, Dict, List, Union
+
+import orjson
+
+from haystack import Pipeline, component
+from haystack.evaluation.eval_utils import (
+    convert_component_outputs_from_dict,
+    convert_component_outputs_to_dict,
+    convert_pipeline_outputs_from_dict,
+    convert_pipeline_outputs_to_dict,
+)
+
+
+class EvaluationResult:
+    """
+    EvaluationResult keeps track of all the information related to evaluation, namely the runnable (Pipeline or component), inputs, outputs, and expected outputs.
+    The EvaluationResult keeps track of all the information stored by eval.
+
+    :param runnable: The runnable (Pipeline or component) used for evaluation.
+    :param inputs: List of inputs used for evaluation.
+    :param outputs: List of outputs generated by the runnable.
+    :param expected_outputs: List of expected outputs used for evaluation.
+    """
+
+    def __init__(
+        self,
+        runnable: Union[Pipeline, component],
+        inputs: List[Dict[str, Any]],
+        outputs: List[Dict[str, Any]],
+        expected_outputs: List[Dict[str, Any]],
+    ) -> None:
+        self.runnable = runnable
+        self.inputs = inputs
+        self.outputs = outputs
+        self.expected_outputs = expected_outputs
+
+    def serialize(self) -> bytes:
+        """
+        Serializes the contents of EvaluationResult, to bytes using orjson:
+        - runnable
+        - inputs
+        - outputs
+        - expected outputs
+
+        The method uses theconvert_objects_to_dict() method to convert the outputs and expected_outputs to serialized data from dataclasses.
+
+        :return bytes: The seriaized data in JSON.
+        """
+        type_runnable = "Pipeline" if isinstance(self.runnable, Pipeline) else "component"
+        if type_runnable == "Pipeline":
+            runnable_type_dict = ("Pipeline", self.runnable.dumps())
+            outputs_dict = convert_pipeline_outputs_to_dict(self.outputs)
+            expected_outputs_dict = convert_pipeline_outputs_to_dict(self.expected_outputs)
+
+        else:
+            runnable_type_dict = ("Component", self.runnable.to_dict())
+            outputs_dict = convert_component_outputs_to_dict(self.outputs)
+            expected_outputs_dict = convert_component_outputs_to_dict(self.expected_outputs)
+
+        final_dict = {
+            "runnable": runnable_type_dict,
+            "inputs": self.inputs,
+            "outputs": outputs_dict,
+            "expected_outputs": expected_outputs_dict,
+        }
+
+        serialized_data = orjson.dumps(final_dict)
+
+        return serialized_data
+
+    @classmethod
+    def deserialize(cls, data: bytes) -> "EvaluationResult":
```

Comment:
I would keep the serialisation methods in line with the other serializable classes, so I'd use `to_dict()` and `from_dict()` instead of `serialize()` and `deserialize()`.

Probably for now it's best if we keep serialisation out of the way and focus on implementing the evaluation only. Later on we can figure out a good strategy to easily save and load evaluation results from file.